src/components/molecules/UserProfilePicture/UserProfilePicture.tsx

```diff
@@ -16,154 +17,176 @@ import {
-          alt={user.anonMode ? DEFAULT_PARTY_NAME : user.partyName}
-        />
-        <S.Avatar
-          onClick={() => openUserProfileModal(user)}
-          className={avatarClassName}
-          backgroundImage={pictureUrl}
-          style={{ ...avatarStyle }}
-        />
+  const openProfileModal = useCallback(() => openUserProfileModal(user), [
+    openUserProfileModal,
+    user,
+  ]);
+
+  // // TODO: extract this logic into a generic custom component + improve it?
+  // const imageErrorHandler = useCallback(
+  //   (event: React.SyntheticEvent<HTMLImageElement, Event>) => {
+  //     const randomAvatar = randomAvatarUrl(user.id);
+  //     setPictureUrl(randomAvatar);
+  //
+  //     // TODO: this isn't very reacty.. is it even needed..?
+  //     // event.currentTarget.onerror = null;
+  //     // event.currentTarget.src = randomAvatar;
+  //   },
+  //   [user.id]
+  // );
+
+  const containerClasses = classNames(
+    "UserProfilePicture",
+    `UserProfilePicture--reaction-${reactionPosition}`
+  );
+
+  const avatarClasses = classNames(
+    "UserProfilePicture__avatar",
+    avatarClassName
+  );
 
+  const avatarStyles = useMemo(
+    () => ({
+      backgroundImage: `url(${pictureUrl})`,
+      ...avatarStyle,
+    }),
+    [avatarStyle, pictureUrl]
+  );
+
+  const userDisplayName = user.anonMode ? DEFAULT_PARTY_NAME : user.partyName;
+
+  return (
+    <div className={containerClasses} style={containerStyle}>
+      {/* TODO: extract this logic into a generic custom component + improve it? Or just remove it entirely? */}
+      {/* Hidden image, used to handle error if image is not loaded */}
+      {/*<img*/}
+      {/*  src={pictureUrl}*/}
+      {/*  onError={(event) => imageErrorHandler(event)}*/}
+      {/*  hidden*/}
+      {/*  style={{ display: "none" }}*/}
+      {/*  alt={user.anonMode ? DEFAULT_PARTY_NAME : user.partyName}*/}
+      {/*/>*/}
+
+      {/* TODO: can we use src/components/atoms/UserAvatar/UserAvatar.tsx here? Should we? */}
+      <div
+        role="img"
+        aria-label={`${userDisplayName}'s avatar`}
+        className={avatarClasses}
+        style={avatarStyles}
+        onClick={openProfileModal}
+      />
+
+      {/* TODO: refactor this to be more performant and potentially memo it if required */}
+      <div className="UserProfilePicture__reaction-container">
         {Reactions.map(
           (reaction, index) =>
             reactions.find(
               (r) => r.created_by === user.id && r.reaction === reaction.type
             ) && (
-              <div key={index} className="reaction-container">
-                <S.Reaction
+              // TODO: don't use index as a key
+              // TODO: should we extract reactions into their own sub-components?
+              // TODO: is reaction-container even defined in our styles here..?
+              <React.Fragment key={index}>
+                <div
+                  className="UserProfilePicture__reaction"
```

Comment:
[TODO] refactor this to be more performant and potentially memo it if required

Reply:
fixed in https://github.com/sparkletown/sparkle/pull/1323/commits/57a6edbbb517f53372760da5a653d9c9471137ad + https://github.com/sparkletown/sparkle/pull/1323/commits/d198f74f1b3c0ff127e8f811fd9694431157e83e + https://github.com/sparkletown/sparkle/pull/1323/commits/c7de18da0c43f2fa81a524a3a4328fbeccc6b21f